### PR TITLE
feat: active campaign segments support

### LINKED ISSUE #812: Add ActiveCampaign segments support
ActiveCampaign supports segments inside lists, which should be supported here. It should be selected through a dropdown, synced with the update of the selected contacts list.

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [ ] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Implements support for sending newsletters to Active Campaign's segments.

Active Campaign currently supports APIs v1 and v3 and to implement this feature we must use both versions. v1 does not have a method for fetching existing segments whereas v3 does not have a method to create campaigns. To handle that in an organized manner, `api_v1_request()` and `api_v3_request()` methods were implemented.

This ESP has campaigns and messages, a campaign can hold multiple messages and choose which message to send on the scheduling/sending action of the campaign. The current sync action creates a message and a campaign, attaching the message to the campaign. This is problematic because neither API (v1 or v3) has an endpoint for editing a campaign. Once a sync occurs with a specific target (sender and list) it cannot be edited through us.

This PR also changes this behavior to only sync the message while editing, the campaign is only created on sending. This fixes the issue by not having to edit a campaign prior to sending.

Closes #812.

### How to test the changes in this Pull Request:

1. Connect to an existing Active Campaign account that has segments
2. Create a new newsletter and select a list and a segment to send to
3. Send and confirm on the ESP dashboard that it was sent to the selected list and segment

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by dkoo — COMMENTED** [review]
Ran into a few issues testing this:

1. When sending a test or real campaign, the request fails with the following response (replicable on each attempt):

```json
{"code":"newspack_newsletters_active_campaign_api_error","message":"Please send this campaign through the web interface to continue.","data":null}
```

Sometimes, I get this instead:

```json
{"code":"http_request_failed","message":"cURL error 28: Resolving timed out after 5514 milliseconds","data":null}
```

As discussed, this seems to indicate some kind of probationary state with the connected ActiveCampaign account, but I'm noting here because I'm not able to bypass it after testing with two different test accounts.

2. When creating a new newsletter, I see some PHP warnings:

```
[09-Jun-2022 21:01:08 UTC] PHP Warning:  Cannot modify header information - headers already sent by (output started at /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:436) in /srv/users/user814b8470/apps/user814b8470/public/wp-admin/admin-header.php on line 9
[09-Jun-2022 21:01:08 UTC] PHP Warning:  Cannot modify header information - headers already sent by (output started at /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:436) in /srv/users/user814b8470/apps/user814b8470/public/wp-includes/option.php on line 1112
[09-Jun-2022 21:01:08 UTC] PHP Warning:  Cannot modify header information - headers already sent by (output started at /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:436) in /srv/users/user814b8470/apps/user814b8470/public/wp-includes/option.php on line 1113
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
[09-Jun-2022 21:01:09 UTC] PHP Notice:  Undefined property: stdClass::$title in /srv/users/user814b8470/apps/user814b8470/public/wp-content/plugins/newspack-newsletters/includes/class-newspack-newsletters-layouts.php on line 214
```

3. When creating a new newsletter for the first time, the select dropdowns in the Newsletter sidebar look very stretched out. Refreshing the editor or editing an existing newsletter fixes the issue, so it only seems to happen when loading this sidebar immediately after creating a new newsletter.

Screen Shot 2022-06-09 at 4 14 37 PM

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php [hunk 1]
```diff
@@ -272,11 +326,11 @@ public function sync( $post ) {
 				__( 'ActiveCampaign API credentials are missing.', 'newspack-newsletters' )
 			);
 		}
-		$campaign_id = get_post_meta( $post->ID, 'ac_campaign_id', true );
-		$from_name   = get_post_meta( $post->ID, 'ac_from_name', true );
-		$from_email  = get_post_meta( $post->ID, 'ac_from_email', true );
-		$list_id     = get_post_meta( $post->ID, 'ac_list_id', true );
-		$is_public   = get_post_meta( $post->ID, 'is_public', true );
+		$from_name  = get_post_meta( $post->ID, 'ac_from_name', true );
```
[the remark this answers is not in the document]
Comment (dkoo) [reply includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:329]: Correction! This also happens when creating a new post of _any_ post type, not just newsletters. Does not seem to happen when editing existing non-newsletter post types, though.

**Comment by miguelpeixe** [comment]
These errors were because of a reminiscent action hook that should be deleted. ae3cca887c5671a10dd5e2b6cd823f24a9a278d9 deletes it.

b8462ee9eaf9442d31c9dc816f55042cfb89513a also improves the trash action, to also remove the created `message` object.

**Comment by miguelpeixe** [comment]
66dc7809f9d9b99dfe43649671896a360f88ffef refactors a few things and fixes the handling of test emails using a disposable campaign. Note that if your account is under internal probation/evaluation on ActiveCampaign's side, the sending of the test might error.

The helper methods `create_campaign()` and `delete_campaign()` abstract the logic for easier handling of the disposable aspect of a campaign. If a campaign fails to send and will be sent again, the previously created campaign is deleted and a new one replaces it to make sure any changes in campaign settings are accounted for.

A few other bugfixes on this commit:

`$campaign_data` for `create_campaign()`:
 - the `p[ $list_id ]` requires the value to also be the `$list_id` and not truesy, so `p [ $list_id ] => $list_id`
 - the `m[ $message_id ]` value refers to the percentage of segmented contacts will receive that message. Defaulting to `100` because this is meant for a campaign with multiple assigned messages.

**Review by dkoo — CHANGES_REQUESTED** [review]
Love the refactoring and the new helper methods to create/delete campaigns. I tried sending another test campaign to myself and this time saw a new error message:

Screen Shot 2022-06-15 at 12 43 43 PM

I also noticed that in ActiveCampaign, the disposable campaign for the test remained. We should probably try to clean up these campaigns even if the test send fails.

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php [hunk 2]
```diff
@@ -227,23 +281,32 @@ public function test( $post_id, $emails ) {
...
 					'action'     => 'test',
-					'campaignid' => $sync_result['campaign_id'],
+					'campaignid' => $campaign['id'],
 					'messageid'  => $sync_result['message_id'],
 					'email'      => implode( ',', $emails ),
 				],
 			]
 		);
+		/** Delete the disposable campaign. */
+		$delete_res = $this->delete_campaign( $campaing['id'], true );
```
Comment (dkoo) [inline includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:306]:
```suggestion
		$delete_res = $this->delete_campaign( $campaign['id'], true );
```

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php [hunk 3]
```diff
@@ -227,23 +281,32 @@ public function test( $post_id, $emails ) {
...
+		$test_result = $this->api_v1_request(
 			'campaign_send',
 			'GET',
 			[
 				'query' => [
 					'type'       => 'html',
 					'action'     => 'test',
-					'campaignid' => $sync_result['campaign_id'],
+					'campaignid' => $campaign['id'],
 					'messageid'  => $sync_result['message_id'],
```
Comment (dkoo) [inline includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:300]: `$sync_result` is not defined in this method, resulting in a warning:

```
PHP Notice:  Undefined variable: sync_result in /srv/users/user92978b3e/apps/user92978b3e/public/wp-content/plugins/newspack-newsletters/includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php on line 300
```

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php [hunk 4]
```diff
@@ -227,23 +281,32 @@ public function test( $post_id, $emails ) {
...
 					'email'      => implode( ',', $emails ),
 				],
 			]
 		);
+		/** Delete the disposable campaign. */
+		$delete_res = $this->delete_campaign( $campaing['id'], true );
+		if ( ! is_wp_error( $delete_res ) ) {
+			delete_post_meta( $post_id, 'ac_test_campaign', $campaign['id'] );
+		}
 		return $test_result;
```
Comment (dkoo) [inline includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:310]: This `test` method returns a response in a different format compared to the other ESP `test` methods. Those other methods return a success message that is shown to the user in the editor UI:

Screen Shot 2022-06-23 at 11 33 17 AM

Can we show a similar message upon successful test send for ActiveCampaign, too?

**Review by dkoo — CHANGES_REQUESTED** [review]
Testing is working for me now! I like the use of the temporary meta values to store the test IDs for async cleanup. There are just a few small things left to fix.

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php (hunk 2 above)
Comment (miguelpeixe) [reply includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:306]: Ha, I think that typo was the issue. 🤦 

Well, no harm in securing the test campaign deletion.

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php (hunk 4 above)
Comment (miguelpeixe) [reply includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php]: Good call! Done in c1590e43361e2fe03f33efaef933761018cedd1c

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php (hunk 3 above)
Comment (miguelpeixe) [reply includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:304]: Nice catch! Fixed by syncing before send in 2c1322e5a1c1e8fb2e79327cb8250b61f7616046

**Review by dkoo — APPROVED** [review]
Latest changes look good! Noting that domain mapping on the AC side looks iffy when sending test campaigns, but that doesn't seem related to the implementation in this plugin.

**Pull request merged**